ratelimits/source_redis.go

```diff
@@ -104,6 +109,43 @@ func (r *RedisSource) Get(ctx context.Context, bucketKey string) (time.Time, err
 	return time.Unix(0, tatNano).UTC(), nil
 }
 
+// BatchGet retrieves the TATs at the specified bucketKeys using a pipelined
+// Redis Transaction in order to reduce the number of round-trips to each Redis
+// shard. An error is returned if the operation failed and nil otherwise. If a
+// bucketKey does not exist, it WILL NOT be included in the returned map.
+func (r *RedisSource) BatchGet(ctx context.Context, bucketKeys []string) (map[string]time.Time, error) {
+	start := r.clk.Now()
+
+	pipeline := r.client.Pipeline()
+	for _, bucketKey := range bucketKeys {
+		pipeline.Get(ctx, bucketKey)
+	}
+	results, err := pipeline.Exec(ctx)
+	if err != nil {
+		r.latency.With(prometheus.Labels{"call": "batchget", "result": resultForError(err)}).Observe(time.Since(start).Seconds())
+		if !errors.Is(err, redis.Nil) {
+			return nil, err
+		}
+	}
+
+	tats := make(map[string]time.Time, len(bucketKeys))
+	for i, result := range results {
+		tatNano, err := result.(*redis.StringCmd).Int64()
```

Comment:
Is each result documented to definitely be of type `*redis.StringCmd`? If not, this should use the `, ok` type assertion and return error if it's the wrong type.

Reply:
Correct. Specifically, each result is documented to return the type for the command that was queued in the Pipline:

A few lines above we queue multiple `Get()` commands:
```go
	pipeline := r.client.Pipeline()
	for _, bucketKey := range bucketKeys {
		pipeline.Get(ctx, bucketKey)
	}
```

The function signature of the `Get()` command:
```go
func (redis.StringCmdable).Get(ctx context.Context, key string) *redis.StringCmd
```